server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/HighlightUtils.java

```diff
@@ -46,24 +48,17 @@ private HighlightUtils() {
      * Load field values for highlighting.
      */
     public static List<Object> loadFieldValues(MappedFieldType fieldType,
+                                               QueryShardContext qsc,
                                                FetchSubPhase.HitContext hitContext,
-                                               boolean forceSource) throws IOException {
-        //percolator needs to always load from source, thus it sets the global force source to true
-        List<Object> textsToHighlight;
-        if (forceSource == false && fieldType.isStored()) {
+                                               boolean storedFieldsAvailable) throws IOException {
+        if (storedFieldsAvailable && fieldType.isStored()) {
             CustomFieldsVisitor fieldVisitor = new CustomFieldsVisitor(singleton(fieldType.name()), false);
             hitContext.reader().document(hitContext.docId(), fieldVisitor);
-            textsToHighlight = fieldVisitor.fields().get(fieldType.name());
-            if (textsToHighlight == null) {
-                // Can happen if the document doesn't have the field to highlight
-                textsToHighlight = Collections.emptyList();
-            }
-        } else {
-            SourceLookup sourceLookup = hitContext.sourceLookup();
-            textsToHighlight = sourceLookup.extractRawValues(fieldType.name());
+            List<Object> textsToHighlight = fieldVisitor.fields().get(fieldType.name());
+            return Objects.requireNonNullElse(textsToHighlight, Collections.emptyList());
         }
-        assert textsToHighlight != null;
-        return textsToHighlight;
+        ValueFetcher fetcher = fieldType.valueFetcher(qsc, null, null);
```

Comment:
I just opened #65292 -- if we like the direction and it gets merged, then `MappedFieldType#valueFetcher` won't need a dedicated `SearchLookup` at all. I think highlighting on runtime fields would 'just work' here (although it'd be worth a test!)